### ISSUE TITLE
Nothing gets logged

### ISSUE THREAD

**Issue by mmoser18** [jsslkeylog/jsslkeylog#8]
Hi - I am trying to use jSSLKeyLog to analyze an issue that I have with a self-written Java application communicating with a SOAP-service  via TLS1.3. I want to decode that connection using WireShark (v4.0.3). I am using Java v17 (downloaded from Eclipse Temurin).

I have downloaded and extracted the jSSLKeyLog.jar file and placed it next to my application's .jar. I then modified the .cmd file that starts my application to  read
`"%JAVA_HOME%\bin\java" -javaagent:jSSLKeyLog.jar=%USERPROFILE%\Desktop\myapp_logfile.log -jar myapp.jar`

That logfile gets indeed created (which suggests to me that jSSLKeyLog gets picked up and started) and my application is communicating with said SOAP service but all that the log file ever receives is the single line

`# SSL/TLS secrets log file, generated by jSSLKeyLog
`

I also tried the "verbose" option (using "==") but that only caused an additional timestamp to appear in the file, not more:

```
## 2023-01-19 17:29:44.130
# SSL/TLS secrets log file, generated by jSSLKeyLog
```

Any idea why there is no more content appearing in this file? Am I missing any further Java options or such?
Any help in getting this working would be highly appreciated!

**Comment by schierlm** [comment]
Hello,

from your report I am unsure whether you downloaded the latest released version (1.3) from the homepage, or the latest master branch from this github repository.

Version 1.3 only supports Java version up to 15. If you want to use any more recent Java version (up to Java 19), you'll currently have to download the source code from this repository and compile it yourself until I find time to spin and upload a new release.

Hope this helps.

**Comment by mmoser18** [comment]
Hi Michael 
Thanks for responding and clarifying this. 
I had indeed downloaded version 1.3. from https://jsslkeylog.github.io/.

I'll see whether I find the time to download and rebuild the project.
Cheers!

**Comment by mmoser18** [comment]
I checked out the project and rebuilt it.

All I had to do was to update the compiler version and - due to that - also the version of the shade plugin (the original version gave me an error that it didn't understand the new class file version).

```
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-compiler-plugin</artifactId>
				<version>3.6.0</version>
				<configuration>
					<source>**17**</source>
					<target>**17**</target>
				</configuration>
			</plugin>
```

and

```
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-shade-plugin</artifactId>
				<version>**3.4.1**</version>
				...
```

With these two changes the plugin worked, wrote to the log file and WireShark was then able to decode the traffic.

I tried to push the change back to the project as a PR but I failed to create a new branch in the project (it wasn't happy with my Github credentials - guess I would need some additional permissions to write to that project...).

HTH and regards,
Michael

**Comment by schierlm** [comment]
It should work with Java 17 even if you do not change the target version (changing the target version will make it **not** work on older versions). What error do you see when you take latest master and just recompile it as is?

Regards,

another Michael :)

**Comment by mmoser18** [comment]
Indeed, it also works without changing the target version! Thanks for pointing that out!

But the shade plugin version still had to be adapted or else I would get "Unsupported class file major version 61" during the build.

So, only the second change above - the maven-shade-plugin version update - is actually required in the pom.xml.
I could of course provide the newly built .jar, if you are interested. But I guess it's even easier and faster for you to do that yourself.

Regards,
Michael

**Comment by schierlm** [comment]
Can you share your Maven version and/or envirronment, and which files exactly have class file version 61 if you (clean and) compile with Java 17, but target 1.7?

I just tried compiling latest master branch with Maven 3.5.2 on Windows 10 using
```
openjdk 17 2021-09-14
OpenJDK Runtime Environment (build 17+35-2724)
```
and the build succeeded and the shaded jar seems to work fine.

**Comment by schierlm** [comment]
I just pushed f5595cdb149214ed7fe5c99141eda6660dd32461 to support the new bytecode format of Java 21. Building is not possible with Java 21 as Java 21 does not support -target 1.7 any longer (only 1.8 or higher). So my two options are either drop -target 1.7 (and point to 1.3 for those old JDKs) or require an older JDK (tested down to 1.8) to build it.

@mmoser18  What would you prefer for the upcoming release? Also, as you are using more bleeding edge Java versions than me, could you test whether the updated jsslkeylog works fine for you on Java 21? I just did a very short test, and logfiles look sane.

**Comment by mmoser18** [comment]
Sorry, I am currently on vacations and will able to check that only in ~3 weeks.
You may want to ping me then again - in case I forget till then.

**Comment by Anonymous-Coward** [comment]
I have used jsslkeylog with multiple Java apps, and it worked. With Java 17, I needed to recompile it. But now I have the same Java app as before, with the same recompiled version, running on exactly the same Java 17 distro, and not a single key gets logged. I do a tcp dump on the port on which the TLS traffic should happen, and I see TLS traffic in the dump, but not a single key gets logged. Is there a way to get some logs out of the agent?

Running context: Alpine-based image with OpenJDK 17. The command line looks like this:

```
java -javaagent:/tmp/jSSLKeyLog.jar=/tmp/ssl.log \
    -Dlogging.level.root=DEBUG \
    -Xms1g -Xmx1g -XX:+ExitOnOutOfMemoryError \
    -cp .:BOOT-INF/lib/*:BOOT-INF/classes the.package.MainClass
```

**Comment by schierlm** [comment]
> But now I have the same Java app as before, with the same recompiled version, running on exactly the same Java 17 distro, and not a single key gets logged.

So what did you change? :) Did your patchlevel of Java 17 change?

Most likely, if it is indeed the same patchlevel, then the application does not use Java for its TLS traffic (so either a native library like APR or it spawns another process).

> Is there a way to get some logs out of the agent?
Apart from the verbose mode (double the equals sign or give `:detailed` option), there is no more logging. You can of course attach a Java debugger to the process and set breakpoints at the places instrumented by the agent. But if the breakpoint never hits, you don't know why either.

**Comment by schierlm** [comment]
Closing for lack of feedback
